query/src/main/java/io/keen/client/java/KeenQueryClient.java

```diff
@@ -0,0 +1,704 @@
+                    for (String key : inputMap.keySet()) {
+                        if (key.equals(KeenQueryConstants.RESULT)) {
+                            // there should not be intervals nested inside GroupByResult's; only
+                            // the other way around.
+                            result = constructQueryResult(inputMap.get(key), false, false);
+                        } else {
+                            properties.put(key, inputMap.get(key));
+                        }
+                    }
+
+                    Group groupBy = new Group(properties);
+                    groupByResult.put(groupBy, result);
+                } else {
+                    throw new IllegalStateException();
+                }
+            } else {
+                throw new IllegalStateException();
+            }
+        }
+
+        return new GroupByResult(groupByResult);
+    }
+
+    /**
+     * Posts a query to the server.
+     *
+     * @param params         The {@link Query} with all the required and optional arguments.
+     * @return The response from the server in the "result" map.
+     * @throws IOException If there was an error communicating with the server or
+     * an error message received from the server.
+     */
+    private Object executeHelper(Query params, Timeframe timeframe) throws IOException {
+        QueryType queryType = params.getQueryType();
+
+        if (false == params.AreParamsValid(queryType)) {
+            throw new IllegalArgumentException("Keen Query parameters are insufficient. Please check Query API docs for required arguments.");
+        }
+
+        String urlString = formatBaseURL(QueryType.getQueryType(queryType));
+
+        // Query parameter args.
+        Map<String, Object> allQueryArgs = params.ConstructQueryArgs();
+        if (timeframe != null) {
+            allQueryArgs.putAll(timeframe.constructTimeframeArgs());
+        }
+
+        Object returnVal = "";
+        URL url = new URL(urlString);
+        returnVal = publishObject(project, url, allQueryArgs);
+
+        return returnVal;
+    }
+
+
+    /**
+     * Posts a request to the server in the specified project, using the given URL and request data.
+     * The request data will be serialized into JSON using the client's
+     * {@link io.keen.client.java.KeenJsonHandler}.
+     *
+     * @param project     The project in which the event(s) will be published; this is used to
+     *                    determine the read key to use for authentication.
+     * @param url         The URL to which the POST should be sent.
+     * @param requestData The request data, which will be serialized into JSON and sent in the
+     *                    request body.
+     * @return The response from the server in the "result" map.
+     * @throws IOException If there was an error communicating with the server.
+     */
+    private Object publishObject(KeenProject project, URL url,
```

Comment:
Nit: Since this isn't actually "publishing" anything, the name `publishObject` is somewhat inappropriate. `postRequest` might be more accurate.